test_tools/onnx_pytorch/common/parser.py

```diff
@@ -0,0 +1,234 @@
+
+class CbDictParser:
+    @staticmethod
+    def parse(obj):
+        import numpy as np
+        import json
+        # features is a json payload wrapped in a bytearray
+        features = json.loads(obj[0].decode('utf-8'))
+        label = obj[1]
+        # I think the logs provide 1-indexed values while the train function is expecting 0-indexed
+        label = label - 1
+        from torch import tensor
+        return {'features': Base64Tensor.parse_dict(features), 'label': label, 'cost': obj[2]}
+
+    
+MSG_TYPE_HEADER = 0x55555555
+MSG_TYPE_CHECKPOINT = 0x11111111
+MSG_TYPE_REGULAR = 0xFFFFFFFF
+MSG_TYPE_EOF = 0xAAAAAAAA
+# mostly ripped from the flatbuf parser
+class JoinedLogStreamReader:
+    def __init__(self, buf):
+        self.buf = buf
+        self.offset = 0
+        self.headers = dict()
+        self.parse_header()
+        self.read_file_header()
+
+    def parse_header(self):
+        if self.buf[0:4] != b'VWFB':
+            raise Exception("Invalid file magic")
+
+        self.version = struct.unpack('I', self.buf[4:8])[0]
+        if self.version != 1:
+            raise Exception(f'Unsuported file version {self.version}')
+        self.offset = 8
+
+    def read(self, size):
+        if size == 0:
+            return bytearray([])
+        data = self.buf[self.offset : self.offset + size]
+        self.offset += size
+        return data
+
+    def read_message(self):
+        kind = struct.unpack('I', self.read(4))[0]
+        length = struct.unpack('I', self.read(4))[0]
+        payload = self.read(length)
+        #discard padding
+        self.read(length % 8)
+        return (kind, payload)
+
+    def read_file_header(self):
+        msg = self.read_message()
+        if msg[0] != MSG_TYPE_HEADER:
+            raise f'Missing file header, found message of type {msg[0]} instead'
+
+        header = FileHeader.GetRootAsFileHeader(msg[1], 0)
+        for i in range(header.PropertiesLength()):
+            p = header.Properties(i)
+            self.headers[p.Key().decode('utf-8')] = p.Value().decode('utf-8')
+
+    def checkpoint_info(self):
+        msg = self.read_message()
+        if msg[0] != MSG_TYPE_CHECKPOINT:
+            raise f'Missing checkpoint info, found message type of {msg[0]} instead'
+        return CheckpointInfo.GetRootAsCheckpointInfo(msg[1], 0)
+
+    def messages(self):
+        while True:
+            msg = self.read_message()
+            if msg[0] == MSG_TYPE_EOF:
+                break
+            yield JoinedPayload.GetRootAsJoinedPayload(msg[1], 0)
+        return None
+
+# partially ripped from the flatbuf parser
+class VWFlatbufferParser:
+    # data should be a bytearray
+    def __init__(self, data):
+        self.reader = JoinedLogStreamReader(data)
+        self.current_id = None
+        self.checkpoint_info = self.reader.checkpoint_info()
+        
+        self.gen = self.read_event_series()
+
+    def __iter__(self):
+        return self.gen
+
+    def getString(self, table):
```

Comment:
Is this naming required by the interface?

Reply:
I have no idea, this is ripped directly from the parser script